src/dune/lib_dep.ml

```diff
@@ -1,32 +1,94 @@
 open Stdune
 
 module Select = struct
-  type choice =
-    { required : Lib_name.Set.t
-    ; forbidden : Lib_name.Set.t
-    ; file : string
-    }
+  module Choice = struct
+    type t =
+      { required : Lib_name.Set.t
+      ; forbidden : Lib_name.Set.t
+      ; file : string
+      }
 
-  let dyn_of_choice { required; forbidden; file } =
-    let open Dyn.Encoder in
-    record
-      [ ("required", Lib_name.Set.to_dyn required)
-      ; ("forbidden", Lib_name.Set.to_dyn forbidden)
-      ; ("file", string file)
-      ]
+    let decode ~result_fn =
+      let open Dune_lang.Decoder in
+      enter
+        (let* dune_version = Dune_lang.Syntax.get_exn Stanza.syntax in
+         let+ loc = loc
+         and+ preds, file =
+           until_keyword "->"
+             ~before:
+               (let+ s = string
+                and+ loc = loc in
+                let loc = Some loc in
+                match String.drop_prefix s ~prefix:"!" with
+                | Some s -> Right (Lib_name.of_string_exn ~loc s)
+                | None -> Left (Lib_name.of_string_exn ~loc s))
+             ~after:(located filename)
+         in
+         match file with
+         | None ->
+           User_error.raise ~loc
+             [ Pp.textf "(<[!]libraries>... -> <file>) expected" ]
+         | Some (loc_file, file) ->
+           let () =
+             if dune_version >= (2, 0) then
+               let prefix =
+                 ( match String.lsplit2 result_fn ~on:'.' with
+                 | None -> result_fn
+                 | Some (s, _) -> s )
+                 ^ "."
+               in
+               if not (String.is_prefix file ~prefix) then
```

Comment:
Should we check the suffix as well? For instance to disallow `(select file.ml (... -> file.x.mli))`